### PR TITLE
POC sticky facets API.

### PR DESCRIPTION
**Pull request by afoucret**
This is a draft PR that aims to share my thoughts on a possible sticky facet API and a possible implementation.

First I have added a sticky flag in the facet definition. When enabled, the facet is disjunctive:

```javascript
var searchOptions = {
  "facets": {
    "city" : {"type": "value", "size": 5, "sticky": true},
    "tags" : {"type": "value", "size": 5},
    "live_music" : {"type": "value"}
  }
}
          
client.search("", searchOptions);
```

If the user choose a value for the facet, it should not be added into the `filters` but at the facet level in the currrentValue properties.

When the user have chosen a single value:
```javascript
var searchOptions = {
  "facets": {
    "city" : {"type": "value", "size": 5, "currentValue":  "New York", "sticky": true},
    "tags" : {"type": "value", "size": 5},
    "live_music" : {"type": "value"}
  }
}
          
client.search("", searchOptions);
```
On multiple values:
```javascript
var searchOptions = {
  "facets": {
    "city" : {"type": "value", "size": 5, "currentValue": ["San Francisco", "New York"], "sticky": true},
    "tags" : {"type": "value", "size": 5},
    "live_music" : {"type": "value"}
  }
}
          
client.search("", searchOptions);
```

The tags facets is not sticky. It means filters are applied using an AND operator. You have the same syntax to apply filters on tags:

```javascript
var searchOptions = {
  "facets": {
    "city" : {"type": "value", "size": 5, "sticky": true},
    "tags" : {"type": "value", "size": 5, "currentValue": ["Bars", "Featured"]},
    "live_music" : {"type": "value"}
  }
}
          
client.search("", searchOptions);
```

Code need to be cleansed and tested.

### CODE REVIEW

**Comment by afoucret** [comment]
A more complete example:

```javascript
var searchOptions = {
  "facets": {
    "city" : {"type": "value", "size": 5, "currentValue": ["San Francisco", "New York"], "sticky": true},
    "tags" : {"type": "value", "size": 5, "currentValue": ["Bars", "Featured"]},
    "live_music" : {"type": "value"}
  }
}
          
client.search("", searchOptions);
```


Will search in `San Francisco` OR in `New York` AND only tagged both `Bars` and `Featured`.

**Comment by JasonStoltz** [comment]
@afoucret The `sticky: true` configuration is great, that exactly how I'd imagine it to work.

Why pass the `currentValue` to facets? We can implicitly determine the current value by looking at the filters that are passed on the request. We're now passing it in two places.

**Comment by afoucret** [comment]
@JasonStoltz

IMHO filters applied through the faceting system are very different than ordinary filters and it is worthy to have another method to apply them.

Let imagine the following use case :

> As an admin of a website, you want to create a page listing restaurant tagged as Japanese and allow faceting on the city field. Here is the query:

{
            "filters" : {"tags": "Japanese"},
            "facets": {
              "city" : {"type": "value", "size": 5, "sticky": true}
}
}

> Now if yout want also to allow faceting on the tags fields:

{
            "filters" : {"tags": "Japanese"},
            "facets": {
              "city" : {"type": "value", "size": 5, "sticky": true},
              "tags" : {"type": "value", "size": 10, "sticky": true
}

The tags facet will display some values that are supposed to be filtered. If you are passing the currently selected values to the facet you can differentiate filters that are coming from the base query and those coming from facet usage.

TL;DR : it make the faceting system more expressive and versatile.

**Comment by yakhinvadim** [comment]
@afoucret @JasonStoltz What is the status of this PR? Do we want to merge this before migrating to Elastic?

**Comment by JasonStoltz** [comment]
@yakhinvadim This was a simple POC, it should not be merged. Feel free to close it.
